compiler/rustc_lint/src/late.rs

```diff
@@ -89,6 +89,14 @@ impl<'tcx, T: LateLintPass<'tcx>> hir_visit::Visitor<'tcx> for LateContextAndPas
     }
 
     fn visit_nested_body(&mut self, body_id: hir::BodyId) {
+        // The typeck_result shouldn't be tainted, otherwise it will cause ICE.
+        // If rustdoc is the caller of this function, we shouldn't run typeck_body here.
+        if !self.context.tcx.sess.opts.actually_rustdoc
+            && self.context.tcx.typeck_body(body_id).tainted_by_errors.is_some()
```

Comment:
Considering this can be called unconditionally outside rustdoc, perform it further down where we do `self.context.cached_typeck_results.set(None);` and just always set it to Some unless in rustdoc mode or if tainted (keep returning in the tainted case of course).

The cache still needs to be an option as we also have lints outside of bodies

Reply:
> Let me clarify, does "the method" point typeck_body?

No I meant `maybe_typeck_results`, which currently has an `or_else`, but could now just remove the or_else and trust that it was already filled in